File: 975/975.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/975.py

using System;
using System.Collections.Generic;
using System.Linq;

class Program975
{
    static List<int> PrimesUpTo(int n)
    {
        var sieve = new bool[n + 1];
        for (int x = 0; x <= n; x++) sieve[x] = true;
        if (n >= 0) sieve[0] = false;
        if (n >= 1) sieve[1] = false;
        int r = (int)Math.Sqrt(n);
        for (int p = 2; p <= r; p++)
        {
            if (sieve[p])
            {
                for (int j = p * p; j <= n; j += p)
                    sieve[j] = false;
            }
        }
        var result = new List<int>();
        for (int i = 0; i <= n; i++)
            if (sieve[i]) result.Add(i);
        return result;
    }

    static int GcdInt(int a, int b)
    {
        while (b != 0) { int t = b; b = a % b; a = t; }
        return a;
    }

    static (int, int) NormalizedPair(int num, int den)
    {
        int common = GcdInt(Math.Abs(num), Math.Abs(den));
        return (num / common, den / common);
    }

    static double HeightVal(int a, int b, (int, int) point)
    {
        var (num, den) = point;
        if (num == 0) return 0.0;
        if (num == den) return 1.0;
        double x = (double)num / (double)den;
        double z = 0.5 - (b * Math.Cos(a * Math.PI * x) + a * Math.Cos(b * Math.PI * x)) / (2.0 * (a + b));
        if (z < 0.0 && z > -1e-14) return 0.0;
        if (z > 1.0 && z < 1.0 + 1e-14) return 1.0;
        return z;
    }

    static int DerivativeIntervalSign(int a, int b, (int, int) left, (int, int) right)
    {
        var (leftNum, leftDen) = left;
        var (rightNum, rightDen) = right;
        double x = (leftNum * rightDen + rightNum * leftDen) / (2.0 * leftDen * rightDen);
        double value = Math.Sin((a + b) * Math.PI * x * 0.5) * Math.Cos(Math.Abs(a - b) * Math.PI * x * 0.5);
        return value > 0.0 ? 1 : -1;
    }

    static Dictionary<(int, int), double[]> turningValuesCache = new Dictionary<(int, int), double[]>();

    static double[] TurningValues(int a, int b)
    {
        var cacheKey = (a, b);
        if (turningValuesCache.ContainsKey(cacheKey)) return turningValuesCache[cacheKey];

        if (a <= 0 || b <= 0 || (a & 1) == 0 || (b & 1) == 0)
            throw new ArgumentException("a,b must be positive odd integers");
        if (a == b) throw new ArgumentException("a != b required");
        int s = a + b;
        int delta = Math.Abs(a - b);
        if (s % 2 != 0 || delta % 2 != 0)
            throw new ArgumentException("For odd a,b, a+b and |a-b| must be even");

        var candidates = new List<(int, int)>();
        for (int k = 0; k <= s / 2; k++)
            candidates.Add(NormalizedPair(2 * k, s));
        for (int k = 0; k < delta / 2; k++)
            candidates.Add(NormalizedPair(2 * k + 1, delta));

        var pointSet = new Dictionary<(int, int), (int, int)>();
        foreach (var p in candidates)
            if (!pointSet.ContainsKey(p)) pointSet[p] = p;

        var points = pointSet.Values.OrderBy(p => (double)p.Item1 / (double)p.Item2).ToList();

        var intervalSigns = new List<int>();
        for (int idx = 0; idx < points.Count - 1; idx++)
            intervalSigns.Add(DerivativeIntervalSign(a, b, points[idx], points[idx + 1]));

        var kept = new List<(int, int)>();
        kept.Add(points[0]);
        for (int idx = 1; idx < points.Count - 1; idx++)
        {
            if (intervalSigns[idx - 1] != intervalSigns[idx])
                kept.Add(points[idx]);
        }
        kept.Add(points[points.Count - 1]);

        var result = new double[kept.Count];
        for (int idx = 0; idx < kept.Count; idx++)
            result[idx] = HeightVal(a, b, kept[idx]);

        turningValuesCache[cacheKey] = result;
        return result;
    }

    static double ComputeF(int a, int b, int c, int d)
    {
        var za = TurningValues(a, b);
        var zb = TurningValues(c, d);
        int i = 0, j = 0;
        double current = 0.0, total = 0.0;
        bool upward = true;
        double eps = 1e-12;
        int maxSteps = 4 * (za.Length + zb.Length) * (za.Length + zb.Length);

        for (int step = 0; step < maxSteps; step++)
        {
            if (i < 0 || i >= za.Length - 1 || j < 0 || j >= zb.Length - 1)
            {
                if (Math.Abs(current - 1.0) < 1e-9) return total;
                throw new Exception("Winding walk left the valid segment range");
            }
            double a0 = za[i], a1 = za[i + 1];
            double b0 = zb[j], b1 = zb[j + 1];
            double lower = Math.Max(Math.Min(a0, a1), Math.Min(b0, b1));
            double upper = Math.Min(Math.Max(a0, a1), Math.Max(b0, b1));
            double nxt = upward ? upper : lower;
            total += Math.Abs(nxt - current);
            bool advanced = false;
            if (Math.Abs(nxt - a0) <= eps) { i -= 1; advanced = true; }
            else if (Math.Abs(nxt - a1) <= eps) { i += 1; advanced = true; }
            if (Math.Abs(nxt - b0) <= eps) { j -= 1; advanced = true; }
            else if (Math.Abs(nxt - b1) <= eps) { j += 1; advanced = true; }
            if (!advanced) throw new Exception("Winding walk did not hit a segment endpoint");
            current = nxt;
            upward = !upward;
        }
        throw new Exception("Exceeded maximum winding-walk steps");
    }

    static double ComputeG(int m, int n)
    {
        var ps = PrimesUpTo(n).Where(p => p >= m).ToList();
        double total = 0.0;
        for (int i = 0; i < ps.Count; i++)
        {
            int p = ps[i];
            for (int k = i + 1; k < ps.Count; k++)
            {
                int q = ps[k];
                total += ComputeF(p, q, p, 2 * q - p);
            }
        }
        return total;
    }

    static void Main(string[] args)
    {
        if (Math.Abs(ComputeF(3, 5, 3, 7) - 7.01772) >= 1e-5) throw new Exception("Assert F(3,5,3,7) failed");
        if (Math.Abs(ComputeF(7, 17, 9, 19) - 26.79578) >= 1e-5) throw new Exception("Assert F(7,17,9,19) failed");
        if (Math.Abs(ComputeG(3, 20) - 463.80866) >= 1e-5) throw new Exception("Assert G(3,20) failed");
        double ans = ComputeG(500, 1000);
        Console.WriteLine(ans.ToString("F5"));
    }
}

File: 975/975.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/975.py

use std::collections::HashMap;
use std::f64::consts::PI;

fn primes_up_to(n: usize) -> Vec<usize> {
    let mut sieve = vec![true; n + 1];
    if n >= 1 {
        sieve[0] = false;
        sieve[1] = false;
    } else if n == 0 {
        sieve[0] = false;
    }
    let r = (n as f64).sqrt() as usize;
    for p in 2..=r {
        if sieve[p] {
            let mut j = p * p;
            while j <= n {
                sieve[j] = false;
                j += p;
            }
        }
    }
    (0..=n).filter(|&i| sieve[i]).collect()
}

fn gcd_int(mut a: i64, mut b: i64) -> i64 {
    while b != 0 {
        let t = b;
        b = a % b;
        a = t;
    }
    a
}

fn normalized_pair(num: i64, den: i64) -> (i64, i64) {
    let common = gcd_int(num.abs(), den.abs());
    (num / common, den / common)
}

fn height_val(a: i64, b: i64, point: (i64, i64)) -> f64 {
    let (num, den) = point;
    if num == 0 {
        return 0.0;
    }
    if num == den {
        return 1.0;
    }
    let x = num as f64 / den as f64;
    let z = 0.5
        - (b as f64 * (a as f64 * PI * x).cos() + a as f64 * (b as f64 * PI * x).cos())
            / (2.0 * (a + b) as f64);
    if z < 0.0 && z > -1e-14 {
        return 0.0;
    }
    if z > 1.0 && z < 1.0 + 1e-14 {
        return 1.0;
    }
    z
}

fn derivative_interval_sign(a: i64, b: i64, left: (i64, i64), right: (i64, i64)) -> i32 {
    let (left_num, left_den) = left;
    let (right_num, right_den) = right;
    let x = (left_num * right_den + right_num * left_den) as f64
        / (2.0 * left_den as f64 * right_den as f64);
    let value = ((a + b) as f64 * PI * x * 0.5).sin()
        * ((a - b).abs() as f64 * PI * x * 0.5).cos();
    if value > 0.0 {
        1
    } else {
        -1
    }
}

fn turning_values(
    a: i64,
    b: i64,
    cache: &mut HashMap<(i64, i64), Vec<f64>>,
) -> Vec<f64> {
    if let Some(v) = cache.get(&(a, b)) {
        return v.clone();
    }

    if a <= 0 || b <= 0 || (a & 1) == 0 || (b & 1) == 0 {
        panic!("a,b must be positive odd integers");
    }
    if a == b {
        panic!("a != b required");
    }
    let s = a + b;
    let delta = (a - b).abs();
    if s % 2 != 0 || delta % 2 != 0 {
        panic!("For odd a,b, a+b and |a-b| must be even");
    }

    let mut candidates: Vec<(i64, i64)> = Vec::new();
    for k in 0..=(s / 2) {
        candidates.push(normalized_pair(2 * k, s));
    }
    for k in 0..(delta / 2) {
        candidates.push(normalized_pair(2 * k + 1, delta));
    }

    let mut seen = std::collections::HashSet::new();
    let mut unique_points: Vec<(i64, i64)> = Vec::new();
    for p in &candidates {
        if seen.insert(*p) {
            unique_points.push(*p);
        }
    }
    unique_points.sort_by(|pa, pb| {
        let fa = pa.0 as f64 / pa.1 as f64;
        let fb = pb.0 as f64 / pb.1 as f64;
        fa.partial_cmp(&fb).unwrap()
    });

    let points = unique_points;
    let mut interval_signs: Vec<i32> = Vec::new();
    for idx in 0..points.len() - 1 {
        interval_signs.push(derivative_interval_sign(a, b, points[idx], points[idx + 1]));
    }

    let mut kept: Vec<(i64, i64)> = Vec::new();
    kept.push(points[0]);
    for idx in 1..points.len() - 1 {
        if interval_signs[idx - 1] != interval_signs[idx] {
            kept.push(points[idx]);
        }
    }
    kept.push(points[points.len() - 1]);

    let result: Vec<f64> = kept.iter().map(|&point| height_val(a, b, point)).collect();
    cache.insert((a, b), result.clone());
    result
}

fn compute_f(
    a: i64,
    b: i64,
    c: i64,
    d: i64,
    cache: &mut HashMap<(i64, i64), Vec<f64>>,
) -> f64 {
    let za = turning_values(a, b, cache);
    let zb = turning_values(c, d, cache);
    let mut i: i64 = 0;
    let mut j: i64 = 0;
    let mut current = 0.0f64;
    let mut total = 0.0f64;
    let mut upward = true;
    let eps = 1e-12f64;
    let max_steps = 4 * (za.len() + zb.len()) * (za.len() + zb.len());

    for _ in 0..max_steps {
        if i < 0 || i >= (za.len() as i64) - 1 || j < 0 || j >= (zb.len() as i64) - 1 {
            if (current - 1.0).abs() < 1e-9 {
                return total;
            }
            panic!("Winding walk left the valid segment range");
        }
        let a0 = za[i as usize];
        let a1 = za[(i + 1) as usize];
        let b0 = zb[j as usize];
        let b1 = zb[(j + 1) as usize];
        let lower = f64::max(f64::min(a0, a1), f64::min(b0, b1));
        let upper = f64::min(f64::max(a0, a1), f64::max(b0, b1));
        let nxt = if upward { upper } else { lower };
        total += (nxt - current).abs();
        let mut advanced = false;
        if (nxt - a0).abs() <= eps {
            i -= 1;
            advanced = true;
        } else if (nxt - a1).abs() <= eps {
            i += 1;
            advanced = true;
        }
        if (nxt - b0).abs() <= eps {
            j -= 1;
            advanced = true;
        } else if (nxt - b1).abs() <= eps {
            j += 1;
            advanced = true;
        }
        if !advanced {
            panic!("Winding walk did not hit a segment endpoint");
        }
        current = nxt;
        upward = !upward;
    }
    panic!("Exceeded maximum winding-walk steps");
}

fn compute_g(m: usize, n: usize, cache: &mut HashMap<(i64, i64), Vec<f64>>) -> f64 {
    let ps: Vec<usize> = primes_up_to(n).into_iter().filter(|&p| p >= m).collect();
    let mut total = 0.0f64;
    for i in 0..ps.len() {
        let p = ps[i] as i64;
        for k in (i + 1)..ps.len() {
            let q = ps[k] as i64;
            total += compute_f(p, q, p, 2 * q - p, cache);
        }
    }
    total
}

fn main() {
    let mut cache: HashMap<(i64, i64), Vec<f64>> = HashMap::new();
    assert!((compute_f(3, 5, 3, 7, &mut cache) - 7.01772).abs() < 1e-5);
    assert!((compute_f(7, 17, 9, 19, &mut cache) - 26.79578).abs() < 1e-5);
    assert!((compute_g(3, 20, &mut cache) - 463.80866).abs() < 1e-5);
    let ans = compute_g(500, 1000, &mut cache);
    println!("{:.5}", ans);
}
